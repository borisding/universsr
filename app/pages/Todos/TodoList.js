import React from 'react';
import PropTypes from 'prop-types';
import styles from './TodoList.module.scss';

export default function TodoList({ todos }) {
  const totalCompleted = todos.filter(todo => !!todo.completed).length;
  return (
    <div className={styles.todosContainer}>
      <h3>{`Todo List (${totalCompleted} / ${todos.length})`}</h3>
      <ul>
        {todos.map(todo => (
          <li key={todo.id}>
            <span
              className={!todo.completed ? styles.pending : styles.completed}
            >
              {todo.title} (ID: {todo.id})
            </span>
          </li>
        ))}
      </ul>
    </div>
  );
}

TodoList.propTypes = {
  todos: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      title: PropTypes.string.isRequired,
      completed: PropTypes.bool.isRequired
    })
  ).isRequired
};
