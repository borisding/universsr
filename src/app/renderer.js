import React from 'react';
import DocumentTitle from 'react-document-title';
import serialize from 'serialize-javascript';
import flushChunks from 'webpack-flush-chunks';
import { renderToString } from 'react-dom/server';
import { flushChunkNames } from 'react-universal-component/server';
import { matchRoutes, renderRoutes } from 'react-router-config';
import { StaticRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import storeFactory from './store';
import routes from './routes';

// preload data for matched route
function prefetchBranchData(store, url) {
  try {
    const branch = matchRoutes(routes, url);
    const promises = branch.map(({ route, match }) => {
      const { loadData } = route;
      const { dispatch } = store;

      if (match && match.isExact && loadData) {
        return Array.isArray(loadData)
          ? Promise.all(loadData.map(action => dispatch(action(match))))
          : dispatch(loadData(match));
      }

      return Promise.resolve(null);
    });

    return Promise.all(promises);
  } catch (err) {
    throw err;
  }
}

// export default server renderer and receiving stats
// also, should allow it to be mounted as middleware for production usage
export default function serverRenderer({ clientStats }) {
  return async (req, res, next) => {
    try {
      const context = {};
      const store = storeFactory();

      await prefetchBranchData(store, req.url);

      const appString = renderToString(
        <Provider store={store}>
          <StaticRouter location={req.url} context={context}>
            {renderRoutes(routes)}
          </StaticRouter>
        </Provider>
      );

      const pageTitle = DocumentTitle.rewind();
      const chunksOptions = { chunkNames: flushChunkNames() };
      const { scripts, stylesheets } = flushChunks(clientStats, chunksOptions);

      const preloadedState = serialize(store.getState(), { isJSON: true });
      const { statusCode = 200, redirectUrl } = context;
      const { nonce } = res.locals;

      if ([301, 302].includes(statusCode) && redirectUrl) {
        return res.redirect(statusCode, redirectUrl);
      }

      res.status(statusCode);
      res.render('index', {
        pageTitle,
        appString,
        preloadedState,
        stylesheets,
        scripts,
        nonce
      });
    } catch (err) {
      next(new Error(err));
    }
  };
}
