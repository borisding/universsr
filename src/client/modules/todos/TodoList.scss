@import '../../common/styles/variables';

label {
  cursor: pointer;
  display: inline-block;
}

ul {
  margin: 0;
  padding: 0;

  > li {
    padding-bottom: $base-padding / 3;
    padding-top: $base-padding / 3;
  }

  .done {
    color: $color-black;
    text-decoration: line-through;
  }

  .pending {
    color: $color-blue;
  }
}
