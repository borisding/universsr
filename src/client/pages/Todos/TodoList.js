import React from 'react';
import PropTypes from 'prop-types';
import { Checkbox } from '@client/common/components';
import './TodoList.module.scss';

export default function TodoList({ todos, updateTodo }) {
  const isTruthy = value => value === true;
  const totalDone = todos.filter(todo => isTruthy(todo.done)).length;

  return (
    <div styleName="todos-container">
      <h3>{`My Todos (${totalDone} / ${todos.length})`}</h3>
      <ul>
        {todos.map(record => (
          <li key={record.id}>
            <label htmlFor={record.id}>
              <Checkbox
                id={record.id}
                value={record.id}
                onChange={evt => updateTodo(evt.target)}
                isChecked={isTruthy(record.done)}
              />
              <span styleName={!record.done ? 'pending' : 'done'}>
                {record.todo}
              </span>
            </label>
          </li>
        ))}
      </ul>
    </div>
  );
}

TodoList.propTypes = {
  todos: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      todo: PropTypes.string.isRequired,
      done: PropTypes.bool.isRequired
    })
  ).isRequired,
  updateTodo: PropTypes.func.isRequired
};
