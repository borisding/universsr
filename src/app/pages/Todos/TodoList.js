import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Checkbox } from '@common/components';
import './styles/TodoList.scss';

export default class TodoList extends Component {
  static propTypes = {
    todos: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.string.isRequired,
        todo: PropTypes.string.isRequired,
        done: PropTypes.bool.isRequired
      })
    ).isRequired,
    updateTodo: PropTypes.func.isRequired
  };

  isTruthy(value) {
    return value === true;
  }

  render() {
    const { todos, updateTodo } = this.props;
    const totalDone = todos.filter(todo => this.isTruthy(todo.done)).length;

    return (
      <div styleName="todos-container">
        <h3>{`My Todos (${totalDone} / ${todos.length})`}</h3>
        <ul>
          {todos.map(record => (
            <li key={record.id}>
              <label htmlFor={record.id}>
                <Checkbox
                  id={record.id}
                  value={record.id}
                  onChange={evt => updateTodo(evt.target)}
                  isChecked={this.isTruthy(record.done)}
                />
                <span styleName={!record.done ? 'pending' : 'done'}>
                  {record.todo}
                </span>
              </label>
            </li>
          ))}
        </ul>
      </div>
    );
  }
}
