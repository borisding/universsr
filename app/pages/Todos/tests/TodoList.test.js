import React from 'react';
import renderer from 'react-test-renderer';
import TodoList from '../TodoList';

describe('TodoList', () => {
  it('renders correctly', () => {
    const props = {
      todos: [{ id: 1, title: 'todo value', completed: false }]
    };
    const tree = renderer.create(<TodoList {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
