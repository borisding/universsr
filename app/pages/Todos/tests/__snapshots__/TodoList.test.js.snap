// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList renders correctly 1`] = `
<div
  className="todosContainer"
>
  <h3>
    Todo List (0 / 1)
  </h3>
  <ul>
    <li>
      <span
        className="pending"
      >
        todo value
         (ID: 
        1
        )
      </span>
    </li>
  </ul>
</div>
`;
